src/completion.ml

```diff
@@ -0,0 +1,147 @@
+(** Completion procedure *)
+
+open Terms
+open Basics
+open Extra
+
+(** total strict order *)
+type 'a order = 'a -> 'a -> bool
+
+let gt : 'a order -> 'a -> 'a -> bool = fun ord a b -> ord a b
+
+(** [ge ord] calculates the reflexive closure of [ord] *)
+let ge : 'a order -> 'a -> 'a -> bool = fun ord a b -> a = b || ord a b
+
+(** [ord_lex ord] calculates the lexicographic order corresponding to the
+    alphabetical order [ord] *)
+let ord_lex : 'a order -> ('a list) order = fun ord l1 l2 ->
+  let rec elim_comm l1 l2 = match (l1, l2) with
+    | h1 :: t1, h2 :: t2 -> begin match ord h1 h2 with
+        | false -> if h1 = h2 then elim_comm t1 t2 else assert false
+        | _     -> l1, l2
+      end
+    | _                  -> l1, l2 in
+  match elim_comm l1 l2 with
+  | [], _              -> false
+  | _, []              -> true
+  | h1 :: _, h2 :: _ -> ord h1 h2
+
+let get_sym : term -> sym = fun t -> match unfold t with
+  | Symb (s, _) -> s
+  | _          -> assert false
+
+(** [lpo ord] calculates the lexicographic path ordering corresponding to
+    the strict total order [ord] on the set of all function symbols *)
+let rec lpo : sym order -> term order = fun ord t1 t2 ->
+  let f, args = get_args t1 in
+  let f = get_sym f in
+  if List.exists (fun x -> ge (lpo ord) x t2) args then true
+  else
+    let g, args' = get_args t2 in
+    let g = get_sym g in
+    match ord f g with
+    | true  ->
+        if List.for_all (fun x -> gt (lpo ord) t1 x) args' then true
+        else
+          false
+    | false ->
+        if f = g then
+          List.for_all (fun x -> gt (lpo ord) t1 x) args' &&
+          ord_lex (lpo ord) args args'
+        else false
+
+(** [build_meta_type k] builds the type “∀(x₁:A₁) (x₂:A₂) ⋯ (xk:Ak), B” where
+    “x₁” through “xk” are fresh variables, “Ai = Mi[x₁,⋯,x(i-1)]”, “Mi” is a
+    new metavar of arity “i-1” and type “∀(x₁:A₂) ⋯ (x(i-1):A(i-1), TYPE”. *)
+let build_meta_type : int -> term = fun k ->
+  assert (k>=0);
+  let vs = Bindlib.new_mvar mkfree (Array.make k "x") in
+  let rec build_prod k p =
+    if k = 0 then p
+    else
+      let k = k-1 in
+      let mk_typ = Bindlib.unbox (build_prod k _Type) in
+      let mk = fresh_meta mk_typ k in
+      let tk = _Meta mk (Array.map _Vari (Array.sub vs 0 k)) in
+      let b = Bindlib.bind_var vs.(k) p in
+      let p = Bindlib.box_apply2 (fun a b -> Prod(a,b)) tk b in
+      build_prod k p
+  in
+  let mk_typ = Bindlib.unbox (build_prod k _Type) (*FIXME?*) in
+  let mk = fresh_meta mk_typ k in
+  let tk = _Meta mk (Array.map _Vari vs) in
+  Bindlib.unbox (build_prod k tk)
+
+type rule_ = sym * pp_hint * rule
```

Comment:
1. rule_ is not nice
2. pp_hint doesn't seem useful. Please remove it.

Then, you can do:
```
type rule = sym * Term.rule
```